File: lib/nnf.ml
open Ast

let neg node = UnaryOperator (Not, node)

let rec double_negation node =
  match node with
  | UnaryOperator (Not, UnaryOperator (Not, child)) -> double_negation child
  | UnaryOperator (Not, Boolean b) when b = false -> Boolean true
  | UnaryOperator (Not, child) -> UnaryOperator (Not, double_negation child)
  | Operator (op, l, r) -> Operator (op, double_negation l, double_negation r)
  | _ -> node

let rec elim_equiv_cond node =
  match node with
  | Operator (Equiv, l, r) ->
    let left = Operator (And, l, r) in
    let right = Operator (And, neg l, neg r) in
    Operator (Or, left, right)
  | Operator (Cond, l, r) ->
    Operator (Or, neg l, r)
  | UnaryOperator (Not, child) ->
    UnaryOperator (Not, elim_equiv_cond child)
  | Operator (op, l, r) ->
    Operator (op, elim_equiv_cond l, elim_equiv_cond r)
  | _ -> node

let rec de_morgan node =
  let bam x = neg @@ de_morgan x in
  (* this function recurs and negates  *)
  match node with
  | UnaryOperator (_, Operator (And, l, r)) ->
    Operator (Or, bam l, bam r)
  | UnaryOperator (_, Operator (Or, l, r)) ->
    Operator (And, bam l, bam r)
  | UnaryOperator (_, node) ->
    UnaryOperator (Not, de_morgan node)
  | Operator (op, l, r) ->
    Operator (op, de_morgan l, de_morgan r)
  | _ -> node

let char_of_op op =
  match op with
  | And -> "&"
  | Or -> "|"
  | Not -> "!"
  | Xor -> "^"
  | Cond -> ">"
  | Equiv -> "="
let rec ast_to_rpn node =
  match node with 
  | Operator (op, l, r) -> ast_to_rpn l ^ ast_to_rpn r ^ char_of_op op
  | UnaryOperator (_, c) -> ast_to_rpn c ^ "!"
  | Var v -> Printf.sprintf "%c" v
  | Boolean b -> if b then "1" else "0"
  | Error msg -> raise (Failure msg)

let rec fixpoint f node =
  let node' = f node in
  if node = node' then node else fixpoint f node'

let ast_to_nnf tree =
  fixpoint double_negation tree
  |> fixpoint elim_equiv_cond
  |> fixpoint de_morgan

let nnf formula =
  Printf.printf "%s -> " formula;
  str_to_tree formula
  |> ast_to_nnf
  |> ast_to_rpn
  |> print_endline

File: readysetrust/src/exercises/nnf.rs
use crate::exercises::evaluation::str_to_tree;
use crate::exercises::{Node, Operator};

fn double_negation(node: Node) -> Node {
    match node {
        Node::Negation(child) => {
            match *child {
                Node::Negation(c) => double_negation(*c),
                _ => Node::Negation(Box::new(double_negation(*child))),
            }
        },
        Node::BinaryOp { .. } => 
            Node::map_children(&node, double_negation),
        _ => node.clone(),
    }
}

// condition
// (A ⇒ B) => (¬A ∨ B)
// and equivalenve
// (A ⇔ B) => ((A ⇒ B) ∧ (B ⇒ A))
// so for equivalence we have to replace it by
// A ⇔ B => (¬A ∨ B) ∧ (A ⇒ ¬B)
fn elim_equiv_cond(node: Node) -> Node {
    match node {
        Node::BinaryOp { op: Operator::Cond, left, right } =>
            Node::BinaryOp {
                op: Operator::Or,
                left: Box::new(Node::Negation(left)),
                right,
            },
        Node::BinaryOp { op: Operator::Equal, left, right } => Node::BinaryOp {
            op: Operator::And,
            left: Box::new(Node::BinaryOp {
                op: Operator::Or,
                left: Box::new(Node::Negation(left.clone())),
                right: right.clone(),
            }),
            right: Box::new(Node::BinaryOp {
                op: Operator::Or,
                left,
                right: Box::new(Node::Negation(right)),
            }),
        },
        _ => node.clone(),
    }
}

fn neg_rec(node: Node) -> Node {
    let expanded_node = de_morgan(node);
    Node::Negation(Box::new(expanded_node))
}

// ¬(A ∨ B) ⇔ (¬A ∧ ¬B)
// ¬(A ∧ B) ⇔ (¬A ∨ ¬B)
fn de_morgan(node: Node) -> Node {
    match node {
        Node::Negation(child) => {
            match *child {
                Node::BinaryOp { op: Operator::Or, left, right } => {
                    Node::BinaryOp {
                        op: Operator::And,
                        left: Box::new(neg_rec(*left)),
                        right: Box::new(neg_rec(*right)),
                    }
                },
                Node::BinaryOp { op: Operator::And, left, right } => {
                    Node::BinaryOp {
                        op: Operator::Or,
                        left: Box::new(neg_rec(*left)),
                        right: Box::new(neg_rec(*right)),
                    }
                },
                _ => Node::Negation(Box::new(de_morgan(*child))),
            }
        },
        _ => node,  // No need to clone if you own the value
    }
}

pub fn negation_normal_form(formula: &str) -> String {
    let ast = str_to_tree(String::from(formula));
    let res = de_morgan(double_negation(ast));
    //ast_to_rpn(&res);
    String::from("yo")
    // let nnf = ast_to_nnf(&ast);
    // ast_to_rpn(&nnf)
}

// tests

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_de_morgan() {
        let node = Node::Negation(Box::new(Node::BinaryOp { op: Operator::Or, left: Box::new(Node::Variable('A')), right: Box::new(Node::Variable('B')) }));
        let expected = Node::BinaryOp { op: Operator::And, left: Box::new(Node::Negation(Box::new(Node::Variable('A')))), right: Box::new(Node::Negation(Box::new(Node::Variable('B')))) };
        // assert_eq!(de_morgan(node), expecFAILted);
        let res = de_morgan(node);
        if res == expected {
            println!("OK");
        } else {
            // panic
            println!("Result: {:?}", res);
            panic!("FAILED");
        }
    }

    #[test]
    fn test_neg_rec() {
        let node = Node::BinaryOp {
            op: Operator::And,
            left: Box::new(Node::Negation(Box::new(Node::Variable('A')))),
            right: Box::new(Node::Negation(Box::new(Node::Variable('B')))),
        };
        let expected = Node::BinaryOp {
            op: Operator::Or,
            left: Box::new(Node::Negation(Box::new(Node::Variable('A')))),
            right: Box::new(Node::Negation(Box::new(Node::Variable('B')))),
        };
        assert_eq!(neg_rec(node), expected);
    }

}